cranelift/codegen/src/isa/riscv64/inst_vector.isle

```diff
@@ -372,6 +440,46 @@
   (if-let neg1 (imm5_from_i8 -1))
   (rv_vxor_vi vs2 neg1 mask vstate))
 
+;; Helper for emitting the `vmax.vv` instruction.
+(decl rv_vmax_vv (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vmax_vv vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VmaxVV) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vmax.vx` instruction.
+(decl rv_vmax_vx (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vmax_vx vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VmaxVX) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vmin.vv` instruction.
+(decl rv_vmin_vv (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vmin_vv vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VminVV) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vmin.vx` instruction.
+(decl rv_vmin_vx (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vmin_vx vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VminVX) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vmaxu.vv` instruction.
+(decl rv_vmaxu_vv (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vmaxu_vv vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VmaxuVV) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vmaxu.vx` instruction.
+(decl rv_vmaxu_vx (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vmaxu_vx vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VmaxuVX) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vminu.vv` instruction.
+(decl rv_vminu_vv (Reg Reg VecOpMasking VState) Reg)
+(rule (rv_vminu_vv vs2 vs1 mask vstate)
+  (vec_alu_rrr (VecAluOpRRR.VminuVV) vs2 vs1 mask vstate))
+
+;; Helper for emitting the `vminu.vx` instruction.
+(decl rv_vminu_vx (Reg Reg VecOpMasking VState) Reg)
```

Comment:
Stray thought, but the x64 backend has different types for `Gpr` and `Xmm` which represent a type-level distinction for different classes of registers. I've found that it's actually worked quite well for x64 and it's saved me a few times from mistakes. I realized I should be carefully reading the rules that use `*_vx` instructions to ensure the variable bound by `splat` comes second instead of first by accident (since I think that would compile past ISLE but wouldn't get past debug asserts perhaps in the backend).

Having this sort of distinction though is a pretty large refactoring so definitely not necessary on this PR, but perhaps something to consider if you also find yourself trying to carefully ensure each register goes to the right spot.

Reply:
Yeah, that sounds like a good idea! And I've definitely had that happen to me. Not with `.vx`, but passing an F register into a `.vx` opcode.

It should be fairly easy to start at least with the vector rules and helpers. I'm going to give that a try.